batch/sql/setup-job-groups.sql

```diff
@@ -0,0 +1,149 @@
+  FOREIGN KEY (`batch_id`, `parent_id`) REFERENCES job_groups (`batch_id`, `job_group_id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+CREATE INDEX `job_group_parents_p_id_level` ON `job_group_parents` (`batch_id`, `parent_id`, `level`);
+CREATE INDEX `job_group_parents_jg_id_level` ON `job_group_parents` (`batch_id`, `job_group_id`, `level`);
+
+ALTER TABLE batches ADD COLUMN migrated_batch BOOLEAN NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+
+ALTER TABLE jobs ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+CREATE INDEX jobs_batch_id_job_group_id ON `jobs` (`batch_id`, `job_group_id`);
+# we do not need an additional index for the scheduler with job groups as job group does not matter
+
+ALTER TABLE batch_attributes ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+CREATE INDEX batch_attributes_batch_id_key_value ON `batch_attributes` (`batch_id`, `job_group_id`, `key`, `value`(256));
+CREATE INDEX batch_attributes_batch_id_value ON `batch_attributes` (`batch_id`, `job_group_id`, `value`(256));
+
+ALTER TABLE batches_cancelled ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+
+ALTER TABLE batches_inst_coll_staging ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+CREATE INDEX batches_inst_coll_staging_batch_id_jg_id ON batches_inst_coll_staging (`batch_id`, `job_group_id`);
+
+ALTER TABLE batch_inst_coll_cancellable_resources ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+CREATE INDEX batch_inst_coll_cancellable_resources_jg_id ON `batch_inst_coll_cancellable_resources` (`batch_id`, `job_group_id`);
+
+ALTER TABLE aggregated_batch_resources_v2 ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+ALTER TABLE aggregated_batch_resources_v3 ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+
+ALTER TABLE batches_n_jobs_in_complete_states ADD COLUMN job_group_id INT NOT NULL DEFAULT 0, ALGORITHM=INSTANT;
+
+DELIMITER $$
+
+DROP TRIGGER IF EXISTS batches_after_update $$
+CREATE TRIGGER batches_after_update AFTER UPDATE ON batches
+FOR EACH ROW
+BEGIN
+  IF OLD.migrated_batch = 0 AND NEW.migrated_batch = 1 THEN
+    INSERT INTO job_groups (batch_id, job_group_id, `user`, cancel_after_n_failures, `state`, n_jobs, time_created, time_completed, callback, attributes)
+    VALUES (NEW.id, 0, NEW.`user`, NEW.cancel_after_n_failures, NEW.state, NEW.n_jobs, NEW.time_created, NEW.time_completed, NEW.callback, NEW.attributes);
+
+    INSERT INTO job_group_parents (batch_id, job_group_id, parent_id, `level`)
+    VALUES (NEW.id, 0, 0, 0);
+  ELSE
+    UPDATE job_groups
+    SET `user` = NEW.`user`,
+      cancel_after_n_failures = NEW.cancel_after_n_failures,
+      `state` = NEW.state,
+      n_jobs = NEW.n_jobs,
+      time_created = NEW.time_created,
+      time_completed = NEW.time_completed,
+      callback = NEW.callback,
+      attributes = NEW.attributes
+    WHERE batch_id = NEW.id AND job_group_id = 0;
```

Comment:
For clarification, are we going to need this part of the trigger forever to ensure that the duplicated state between job groups and batches is updated in unison?

Reply:
Do we want to allow things like the `user` to change?

Reply:
No. The `user` is for an index that is needed. It shouldn't change between job groups in the same batch, but I wanted to be complete here that any field updates are reflected in the job groups table.

As for needing the trigger forever, I think we have independent insert and update commands in the batch front end / driver that keep the state in sync and not in the trigger. At some point, I want most of these fields to be essentially deprecated on the batches table. Thoughts??

Reply:
ah, ya I appreciate wanting to be complete here but as a reader I interpret this less as being complete and more as "accounting for all of the fields that could change". Can we instead error if the new value differs from the old one?